File: apps/client/src/canon/extract.ts
/** 浏览器预览用的启发式抽取，与 `novel-story-model::extract` 对齐。 */

const SPEAKER_SUFFIXES = [
  "冷冷道",
  "淡淡道",
  "低声道",
  "缓缓道",
  "说道",
  "问道",
  "笑道",
  "喝道",
  "叹道",
  "答道",
  "怒道",
  "叫道",
];

const LOCATION_PREFIXES = ["来到了", "走进了", "抵达了", "来到", "走进", "抵达"];

const NAME_STOPWORDS = new Set([
  "然后",
  "于是",
  "只是",
  "可是",
  "但是",
  "这时",
  "此时",
  "忽然",
  "突然",
  "终于",
  "其实",
  "因为",
  "所以",
  "如果",
  "虽然",
  "他们",
  "她们",
  "我们",
  "自己",
  "这个",
  "那个",
  "什么",
  "没有",
  "不是",
  "已经",
  "还是",
  "或者",
  "以及",
]);

export interface Mention {
  entityName: string;
  entityKind: "character" | "location" | "item";
  predicate: string;
  object: string;
  quote: string;
  confidence: number;
}

function isCjk(ch: string): boolean {
  const code = ch.codePointAt(0) ?? 0;
  return code >= 0x4e00 && code <= 0x9fff;
}

function precedingCjkName(text: string, at: number): string | null {
  let end = at;
  while (end > 0 && /\s/.test(text[end - 1] ?? "")) end -= 1;
  let start = end;
  let count = 0;
  while (start > 0 && count < 8) {
    const ch = text[start - 1] ?? "";
    if (isCjk(ch)) {
      start -= 1;
      count += 1;
    } else {
      break;
    }
  }
  if (count < 2) return null;
  return text.slice(start, end);
}

function followingCjkName(text: string, from: number, min: number, max: number): string | null {
  let start = from;
  while (start < text.length && /\s/.test(text[start] ?? "")) start += 1;
  let end = start;
  let count = 0;
  while (end < text.length && count < max) {
    const ch = text[end] ?? "";
    if (isCjk(ch)) {
      end += 1;
      count += 1;
    } else {
      break;
    }
  }
  if (count < min) return null;
  return text.slice(start, end);
}

function quoteAround(text: string, start: number, end: number): string {
  return text.slice(Math.max(0, start - 12), Math.min(text.length, end + 12)).replace(/\n/g, " ").trim();
}

export function extractMentions(text: string): Mention[] {
  const mentions: Mention[] = [];

  for (const suffix of SPEAKER_SUFFIXES) {
    let search = 0;
    while (search < text.length) {
      const at = text.indexOf(suffix, search);
      if (at < 0) break;
      const name = precedingCjkName(text, at);
      if (name && !NAME_STOPWORDS.has(name)) {
        const end = at + suffix.length;
        mentions.push({
          entityName: name,
          entityKind: "character",
          predicate: "appearsAsSpeaker",
          object: name,
          quote: quoteAround(text, at - name.length, end),
          confidence: 0.82,
        });
      }
      search = at + suffix.length;
    }
  }

  let search = 0;
  while (search < text.length) {
    const start = text.indexOf("《", search);
    if (start < 0) break;
    const close = text.indexOf("》", start + 1);
    if (close < 0) break;
    const title = text.slice(start + 1, close);
    if (title.length >= 2 && title.length <= 20 && !title.includes("\n")) {
      mentions.push({
        entityName: title,
        entityKind: "item",
        predicate: "titledWork",
        object: title,
        quote: quoteAround(text, start, close + 1),
        confidence: 0.78,
      });
    }
    search = close + 1;
  }

  for (const prefix of LOCATION_PREFIXES) {
    let locSearch = 0;
    while (locSearch < text.length) {
      const at = text.indexOf(prefix, locSearch);
      if (at < 0) break;
      const after = at + prefix.length;
      const name = followingCjkName(text, after, 2, 12);
      if (name && !NAME_STOPWORDS.has(name)) {
        mentions.push({
          entityName: name,
          entityKind: "location",
          predicate: "mentionedLocation",
          object: name,
          quote: quoteAround(text, at, after + name.length),
          confidence: 0.64,
        });
      }
      locSearch = after;
    }
  }

  const seen = new Set<string>();
  return mentions.filter((mention) => {
    const key = `${mention.entityKind}|${mention.entityName}|${mention.predicate}`;
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

File: crates/story-model/src/extract.rs
//! 从正文抽取正史候选。首版只用确定性启发式，不调用模型。
//!
//! 规则（网文常见写法）：
//! - `林晚说道` / `沈雾问道` → 人物
//! - `《潮汐秘录》` → 物/典籍
//! - `走进雾港码头` / `来到青州` → 地点

use novel_domain::{EntityKind, ExtractedMention};
use std::collections::HashSet;

const SPEAKER_SUFFIXES: &[&str] = &[
    "冷冷道",
    "淡淡道",
    "低声道",
    "缓缓道",
    "说道",
    "问道",
    "笑道",
    "喝道",
    "叹道",
    "答道",
    "怒道",
    "叫道",
];

const LOCATION_PREFIXES: &[&str] = &["来到了", "走进了", "抵达了", "来到", "走进", "抵达"];

const NAME_STOPWORDS: &[&str] = &[
    "然后", "于是", "只是", "可是", "但是", "这时", "此时", "忽然", "突然", "终于", "其实", "因为",
    "所以", "如果", "虽然", "他们", "她们", "我们", "自己", "这个", "那个", "什么", "没有", "不是",
    "已经", "还是", "或者", "以及",
];

pub fn extract_mentions(text: &str) -> Vec<ExtractedMention> {
    let mut mentions = Vec::new();
    extract_speakers(text, &mut mentions);
    extract_titles(text, &mut mentions);
    extract_locations(text, &mut mentions);
    dedupe(mentions)
}

fn extract_speakers(text: &str, out: &mut Vec<ExtractedMention>) {
    for suffix in SPEAKER_SUFFIXES {
        let mut search = 0;
        while let Some(rel) = text[search..].find(suffix) {
            let at = search + rel;
            if let Some(name) = preceding_cjk_name(text, at) {
                if is_plausible_name(name) {
                    let end = at + suffix.len();
                    out.push(ExtractedMention {
                        entity_name: name.to_owned(),
                        entity_kind: EntityKind::Character,
                        predicate: "appearsAsSpeaker".into(),
                        object: name.to_owned(),
                        quote: quote_around(text, at.saturating_sub(name.len()), end),
                        confidence: 0.82,
                    });
                }
            }
            search = at + suffix.len();
        }
    }
}

fn extract_titles(text: &str, out: &mut Vec<ExtractedMention>) {
    let mut search = 0;
    while let Some(rel) = text[search..].find('《') {
        let start = search + rel;
        let after = start + '《'.len_utf8();
        if let Some(rel_end) = text[after..].find('》') {
            let title = &text[after..after + rel_end];
            if (2..=20).contains(&title.chars().count()) && !title.contains('\n') {
                let end = after + rel_end + '》'.len_utf8();
                out.push(ExtractedMention {
                    entity_name: title.to_owned(),
                    entity_kind: EntityKind::Item,
                    predicate: "titledWork".into(),
                    object: title.to_owned(),
                    quote: quote_around(text, start, end),
                    confidence: 0.78,
                });
            }
            search = after + rel_end + '》'.len_utf8();
        } else {
            break;
        }
    }
}

fn extract_locations(text: &str, out: &mut Vec<ExtractedMention>) {
    for prefix in LOCATION_PREFIXES {
        let mut search = 0;
        while let Some(rel) = text[search..].find(prefix) {
            let at = search + rel;
            let after = at + prefix.len();
            if let Some(name) = following_cjk_name(text, after, 2, 12) {
                if is_plausible_name(name) {
                    let end = after + name.len();
                    out.push(ExtractedMention {
                        entity_name: name.to_owned(),
                        entity_kind: EntityKind::Location,
                        predicate: "mentionedLocation".into(),
                        object: name.to_owned(),
                        quote: quote_around(text, at, end),
                        confidence: 0.64,
                    });
                }
            }
            search = after;
        }
    }
}

fn preceding_cjk_name(text: &str, suffix_start: usize) -> Option<&str> {
    let mut end = suffix_start;
    while end > 0 {
        let ch = text[..end].chars().next_back()?;
        if ch.is_whitespace() {
            end -= ch.len_utf8();
        } else {
            break;
        }
    }
    let mut start = end;
    let mut count = 0usize;
    while start > 0 && count < 8 {
        let ch = text[..start].chars().next_back()?;
        if is_cjk(ch) {
            start -= ch.len_utf8();
            count += 1;
        } else {
            break;
        }
    }
    if count < 2 {
        return None;
    }
    Some(&text[start..end])
}

fn following_cjk_name(text: &str, from: usize, min: usize, max: usize) -> Option<&str> {
    let mut start = from;
    while start < text.len() {
        let ch = text[start..].chars().next()?;
        if ch.is_whitespace() {
            start += ch.len_utf8();
        } else {
            break;
        }
    }
    let mut end = start;
    let mut count = 0usize;
    while end < text.len() && count < max {
        let ch = text[end..].chars().next()?;
        if is_cjk(ch) {
            end += ch.len_utf8();
            count += 1;
        } else {
            break;
        }
    }
    if count < min {
        return None;
    }
    Some(&text[start..end])
}

fn is_cjk(c: char) -> bool {
    ('\u{4e00}'..='\u{9fff}').contains(&c)
}

fn is_plausible_name(name: &str) -> bool {
    !NAME_STOPWORDS.contains(&name)
}

fn quote_around(text: &str, start: usize, end: usize) -> String {
    let start = floor_char(text, start.saturating_sub(12));
    let end = ceil_char(text, (end + 12).min(text.len()));
    text[start..end].replace('\n', " ").trim().to_owned()
}

fn floor_char(text: &str, mut index: usize) -> usize {
    index = index.min(text.len());
    while index > 0 && !text.is_char_boundary(index) {
        index -= 1;
    }
    index
}

fn ceil_char(text: &str, mut index: usize) -> usize {
    index = index.min(text.len());
    while index < text.len() && !text.is_char_boundary(index) {
        index += 1;
    }
    index
}

fn dedupe(mentions: Vec<ExtractedMention>) -> Vec<ExtractedMention> {
    let mut seen = HashSet::new();
    let mut out = Vec::new();
    for mention in mentions {
        let key = (
            mention.entity_kind.clone(),
            mention.entity_name.clone(),
            mention.predicate.clone(),
        );
        if seen.insert(key) {
            out.push(mention);
        }
    }
    out
}
